pkg/liqo-controller-manager/webhooks/shadowpod/cache.go

```diff
@@ -0,0 +1,269 @@
+	pi.mu.Lock()
+	defer pi.mu.Unlock()
+	klog.V(5).Infof("Found %d ShadowPods for cluster %q", len(shadowPodList.Items), pi.clusterIdentity.String())
+	// Check on all cluster ShadowPods and saving them in a temporary map
+	spMap := pi.checkShadowPodsAlignment(shadowPodList)
+	klog.V(5).Infof("Searching for terminated ShadowPodDescription to be removed from cache")
+	// Alignment of all ShadowPodDescriptions in cache
+	pi.alignTerminatingShadowPodDescriptions(spMap)
+}
+
+func (pi *peeringInfo) alignTerminatingShadowPodDescriptions(spMap map[string]struct{}) {
+	for _, shadowPodDescription := range pi.shadowPods {
+		// Check if the ShadowPod is in terminating phase in cache and has been already terminated/deleted from the cluster
+		// if true ShadowPodDescription can be also deleted from the cache
+		if _, stillPresent := spMap[shadowPodDescription.namespacedName.String()]; !stillPresent {
+			if !shadowPodDescription.running {
+				pi.removeShadowPod(shadowPodDescription)
+				klog.V(5).Infof("ShadowPodDescription %s removed from cache", shadowPodDescription.namespacedName.String())
+			} else if time.Since(shadowPodDescription.creationTimestamp) > 30*time.Second {
+				pi.terminateShadowPod(shadowPodDescription)
+				pi.removeShadowPod(shadowPodDescription)
+			}
+		}
+	}
+}
+
+func (pi *peeringInfo) alignExistingShadowPods(shadowPodList *vkv1alpha1.ShadowPodList) {
+	pi.mu.Lock()
+	defer pi.mu.Unlock()
+	klog.V(5).Infof("Found %d ShadowPods running on cluster %s", len(shadowPodList.Items), pi.clusterIdentity.String())
+	for i := range shadowPodList.Items {
+		shadowPod := &shadowPodList.Items[i]
+		nsname := types.NamespacedName{Name: shadowPod.Name, Namespace: shadowPod.Namespace}
+		_, found := pi.shadowPods[nsname.String()]
+		if !found {
+			// Errors are intentionally ignored here. Pod is already running on cluster.
+			spQuota, _ := getQuotaFromShadowPod(shadowPod, false)
+			pi.addShadowPod(createShadowPodDescription(shadowPod.GetName(), shadowPod.GetNamespace(), shadowPod.GetUID(), spQuota))
+			klog.V(4).Infof("ShadowPod %s added in cache", nsname.String())
+		}
+	}
+}
+
+func (pi *peeringInfo) checkShadowPodsAlignment(shadowPodList *vkv1alpha1.ShadowPodList) map[string]struct{} {
+	spMap := make(map[string]struct{})
+	for i := range shadowPodList.Items {
+		shadowPod := &shadowPodList.Items[i]
+		nsname := types.NamespacedName{Name: shadowPod.Name, Namespace: shadowPod.Namespace}
+		_, found := pi.shadowPods[nsname.String()]
+		if !found {
+			klog.Warningf("Warning: ShadowPod %s not found in cache", nsname.String())
+			spQuota, _ := getQuotaFromShadowPod(shadowPod, false)
+			pi.addShadowPod(createShadowPodDescription(shadowPod.GetName(), shadowPod.GetNamespace(), shadowPod.GetUID(), spQuota))
+			continue
+		}
+		// Saving a list all the ShadowPods in the cluster that are also in cache
+		spMap[nsname.String()] = struct{}{}
+	}
+	return spMap
+}
```

Comment:
This two functions appear to be mostly overlapping. Would consider whether they could be refactored to reduce code duplication.